parachain/pallets/omni-account/src/lib.rs

```diff
@@ -0,0 +1,216 @@
+mod tests;
+
+pub use core_primitives::{Identity, MemberIdentity};
+pub use frame_system::pallet_prelude::BlockNumberFor;
+pub use pallet::*;
+
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+use sp_core::H256;
+use sp_core_hashing::blake2_256;
+use sp_std::vec::Vec;
+
+pub trait AccountIdConverter<T: Config> {
+	fn convert(account_id: T::AccountId) -> Identity;
+}
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	/// The current storage version.
+	const STORAGE_VERSION: StorageVersion = StorageVersion::new(0);
+
+	#[pallet::pallet]
+	#[pallet::storage_version(STORAGE_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The event type of this pallet.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+		/// The origin which can manage the pallet.
+		type TEECallOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+		/// The maximum number of identities an id graph can have.
+		#[pallet::constant]
+		type MaxIDGraphLength: Get<u32>;
+		/// AccountId converter
+		type AccountIdConverter: AccountIdConverter<Self>;
+	}
+	pub type IDGraphLinks<T> = BoundedVec<(H256, MemberIdentity), <T as Config>::MaxIDGraphLength>;
+
+	#[pallet::storage]
+	pub type LinkedIdentityHashes<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = H256, Value = ()>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn id_graphs)]
+	pub type IDGraphs<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = T::AccountId, Value = IDGraphLinks<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn id_graph_hashes)]
+	pub type IDGraphHashes<T: Config> =
+		StorageMap<Hasher = Blake2_128Concat, Key = T::AccountId, Value = H256>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Identity linked
+		IdentityLinked { who: T::AccountId, identity: H256 },
+		/// Identity remove
+		IdentityRemoved { who: T::AccountId, identity_hashes: Vec<H256> },
+		/// Identity made public
+		IdentityMadePublic { who: T::AccountId, identity_hash: H256 },
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Identity is already linked
+		IdentityAlreadyLinked,
+		/// IDGraph len limit reached
+		IDGraphLenLimitReached,
+		/// Identity not found
+		IdentityNotFound,
+		/// Invalid identity
+		PrimeIdentityInvalid,
+		/// Prime identity not found
+		PrimeIdentityNotFound,
+		/// Identity is private
+		IdentityIsPrivate,
+		/// Identities empty
+		IdentitiesEmpty,
+		/// IDGraph hash does not match
+		IDGraphHashMismatch,
+		/// Missing IDGraph hash
+		IDGraphHashMissing,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight((195_000_000, DispatchClass::Normal))]
+		pub fn link_identity(
+			origin: OriginFor<T>,
+			who: T::AccountId,
+			identity_hash: H256,
+			identity: MemberIdentity,
+			maybe_id_graph_hash: Option<H256>,
```

Comment:
I think we need to record both old and new IDGraph:
- old: the current IDGraph
- new: the pre-computed IDGraph if the identity is successfully linked

And how to compute IDGraphHash: maybe the merkle root of all member identity hashes? The hash of encoded IDGraph could change as you make some identity public?

Reply:
Here is the idea of the current implementation: Worker A and Worker B both want want to update the IDGraph, they both have the current `id_graph_hash`, if Worker B updates it first then the hash will change so the call from Worker A will be rejected. Does it makes sense?

> And how to compute IDGraphHash: maybe the merkle root of all member identity hashes? The hash of encoded IDGraph could change as you make some identity public?

This is a good point, using only the identity hashes makes sense

Reply:
To keep it simple I've updated the `id_graph_hash` calculation to be the hash of all the identity_hashes